dnf/dnf-utils.c

```diff
@@ -36,12 +70,76 @@ dnf_utils_print_transaction (DnfContext *ctx)
       return FALSE;
     }
-  g_print ("Transaction: %u packages\n", pkgs->len);
 
-  for (guint i = 0; i < pkgs->len; i++)
+  struct libscols_table *tb;
+  struct libscols_line *ln;
+
+  tb = scols_new_table ();
+  scols_table_new_column (tb, " Package",   0.7, SCOLS_FL_TRUNC);
+  scols_table_new_column (tb, "Repository", 0.2, SCOLS_FL_TRUNC);
+  scols_table_new_column (tb, "Size",       0.1, SCOLS_FL_RIGHT);
+  scols_table_enable_maxout (tb, 1);
+
+  g_autoptr(GPtrArray) pkgs_install = dnf_goal_get_packages (dnf_context_get_goal (ctx),
+                                                             DNF_PACKAGE_INFO_INSTALL,
+                                                             -1);
+  if (pkgs_install->len != 0)
     {
-      DnfPackage *pkg = pkgs->pdata[i];
-      g_print ("%s (%s)\n", dnf_package_get_nevra (pkg), dnf_package_get_reponame (pkg));
+      ln = scols_table_new_line(tb, NULL);
+      scols_line_set_data(ln, COL_NEVRA, "Installing:");
+      dnf_utils_add_transaction_packages (tb, pkgs_install);
+    }
+
+
+  g_autoptr(GPtrArray) pkgs_reinstall = dnf_goal_get_packages (dnf_context_get_goal (ctx),
+                                                               DNF_PACKAGE_INFO_REINSTALL,
+                                                               -1);
+  if (pkgs_reinstall->len != 0)
+    {
+      ln = scols_table_new_line(tb, NULL);
+      scols_line_set_data(ln, COL_NEVRA, "Reinstalling:");
+      dnf_utils_add_transaction_packages (tb, pkgs_reinstall);
+    }
+
+  g_autoptr(GPtrArray) pkgs_upgrade = dnf_goal_get_packages (dnf_context_get_goal (ctx),
+                                                             DNF_PACKAGE_INFO_UPDATE,
+                                                             -1);
+  if (pkgs_upgrade->len != 0)
+    {
+      ln = scols_table_new_line(tb, NULL);
+      scols_line_set_data(ln, COL_NEVRA, "Upgrading:");
+      dnf_utils_add_transaction_packages (tb, pkgs_upgrade);
     }
+
+  g_autoptr(GPtrArray) pkgs_remove = dnf_goal_get_packages (dnf_context_get_goal (ctx),
+                                                            DNF_PACKAGE_INFO_REMOVE,
+                                                            -1);
+  if (pkgs_remove->len != 0)
+    {
+      ln = scols_table_new_line(tb, NULL);
+      scols_line_set_data(ln, COL_NEVRA, "Removing:");
+      dnf_utils_add_transaction_packages (tb, pkgs_remove);
+    }
+
+  g_autoptr(GPtrArray) pkgs_downgrade = dnf_goal_get_packages (dnf_context_get_goal (ctx),
+                                                               DNF_PACKAGE_INFO_DOWNGRADE,
+                                                               -1);
+  if (pkgs_downgrade->len != 0)
+    {
+      ln = scols_table_new_line(tb, NULL);
+      scols_line_set_data(ln, COL_NEVRA, "Downgrading:");
+      dnf_utils_add_transaction_packages (tb, pkgs_downgrade);
+    }
+
+  scols_print_table(tb);
+  scols_unref_table(tb);
+
+  g_print ("Transaction Summary:\n");
```

Comment:
just read code... hmmm, I think it is fine to just print it (but I think it's better to combine it with existing table)